# 1.Simple Imputation Technique

# 1.1.Mean/Median Imputation

# 1.1.1. Mean Imputation

In [47]:
# import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# load the Titanic dataset
df = sns.load_dataset("titanic")
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [48]:
# check the number of missing values in each column
df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

 We can see that the `age` column has 177 missing values. Let's replace these missing values with the mean of the column:

In [49]:
# impute the missing values with mean
df['age'] = df['age'].fillna(df['age'].mean())

# check the number of missing values in each column
df.isnull().sum().sort_values(ascending=False)

deck           688
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

We can see that the missing values in the `age` column have been replaced with the mean of the column.

# 1.1.2 Median Imputation

In [90]:
df = sns.load_dataset('titanic')
df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

In [91]:
# impute the missing values with mean
df['age'] = df['age'].fillna(df['age'].median())

# check the number of missing values in each column
df.isnull().sum().sort_values(ascending=False)

deck           688
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

# 1.2 Mode Imputation

In [92]:
# load the dataset 
df = sns.load_dataset('titanic')

# check the number of missing values in each column
df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

In [93]:
# impute missing values with mode
df['embark_town'] = df['embark_town'].fillna(df['embark_town'].mode()[0])
df['embarked'] = df['embarked'].fillna(df['embarked'].mode()[0])

# check the number of missing values in each column 
df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
embarked         0
class            0
who              0
adult_male       0
embark_town      0
alive            0
alone            0
dtype: int64

# 2. K-Nearest Neigbors (KNN)

In [50]:
# load the dataset
df = sns.load_dataset("titanic")

# check the numbers of missing values in each column
df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

In [51]:
# impute the missing values with KNN imputer 
from sklearn.impute import KNNImputer

# call the KNN class wit number of neighbors = 4
imputer = KNNImputer(n_neighbors=4)

# impute the missing values with KNN imputer 
df['age'] = imputer.fit_transform(df[['age']])

# check the number of missing values in each column
df.isnull().sum().sort_values(ascending=False)

deck           688
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

# 3. Regression Imputation

In [52]:
# load the dataset
df = sns.load_dataset("titanic")

# check the numbers of missing values in each column
df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

In [53]:
# impute missing values with regression imputer 
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

# call the IterativeImputer class with max_iter = 10
imputer = IterativeImputer(max_iter = 10)

# impute missing values with regression imputer 
df['age'] = imputer.fit_transform(df[['age']])

# check the number of missing values in each column
df.isnull().sum().sort_values(ascending=False)

deck           688
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

# 4. Random Forests for Imputing Missing Values

In [54]:
# import libraries 
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, mean_absolute_percentage_error
from sklearn.impute import SimpleImputer

# load the dataset
df = sns.load_dataset("titanic")

# check the numbers of missing values in each column
df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

In [55]:
# remove the deck column
df.drop('deck', axis=1, inplace=True)

# check missing values in each column
df.isnull().sum().sort_values(ascending=False)

age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

We will encode the data at this stage:

In [56]:
df.head()


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male  embark_town alive  alone  
0    man        True  Southampton    no  False  
1  woman       False    Cherbourg   yes  False  
2  woman       False  Southampton   yes   True  
3  woman       False  Southampton   yes  False  
4    man        True  Southampton    no   True

In [57]:
# encode the data using label encoding 
from sklearn.preprocessing import LabelEncoder

# columns to encode 
columns_to_encode = ['sex', 'embarked', 'class', 'who', 'embark_town', 'alive']

# dictionary to store labelencoders for each column
label_encoders = {}

# loop to apply labelencoder to each column
for col in columns_to_encode:
    # create a new labelencoder for each column
    le = LabelEncoder()

    # fit and transform the data, then inverse transform it 
    df[col] = le.fit_transform(df[col])

    # store the encoder in the dictionary 
    label_encoders[col] = le

# check the first few rows of the dataframe
df.head()

survived  pclass  sex   age  sibsp  parch     fare  embarked  class  who  \
0         0       3    1  22.0      1      0   7.2500         2      2    1   
1         1       1    0  38.0      1      0  71.2833         0      0    2   
2         1       3    0  26.0      0      0   7.9250         2      2    2   
3         1       1    0  35.0      1      0  53.1000         2      0    2   
4         0       3    1  35.0      0      0   8.0500         2      2    1   

   adult_male  embark_town  alive  alone  
0        True            2      0  False  
1       False            0      1  False  
2       False            2      1   True  
3       False            2      1  False  
4        True            2      0   True

We have to first impute the missing values in the age column before we can use it to predict the missing values in the `embarked` and `embark_town` columns.

In [58]:
# split the dataset into two parts: one with missing values , one without 
df_with_missing = df[df['age'].isna()]

# dropna removes all rows with missing values
df_without_missing = df.dropna()

Let's see the shape of the datasets with and without the missing values:

In [59]:
print("The shape of the original dataset is:" , df.shape )
print("The shape of the dataset without missing values removed:" , df_without_missing.shape)
print("The shape of the dataset with missing values is:" , df_with_missing.shape)

The shape of the original dataset is: (891, 14)
The shape of the dataset without missing values removed: (714, 14)
The shape of the dataset with missing values is: (177, 14)


let's see the first five rows of the dataset with the missing values

In [60]:
df_with_missing.head()

survived  pclass  sex  age  sibsp  parch     fare  embarked  class  who  \
5          0       3    1  NaN      0      0   8.4583         1      2    1   
17         1       2    1  NaN      0      0  13.0000         2      1    1   
19         1       3    0  NaN      0      0   7.2250         0      2    2   
26         0       3    1  NaN      0      0   7.2250         0      2    1   
28         1       3    0  NaN      0      0   7.8792         1      2    2   

    adult_male  embark_town  alive  alone  
5         True            1      0   True  
17        True            2      1   True  
19       False            0      1   True  
26        True            0      0   True  
28       False            1      1   True

let's see the first five rows of the dataset without the missing values

In [61]:
df_without_missing.head()

survived  pclass  sex   age  sibsp  parch     fare  embarked  class  who  \
0         0       3    1  22.0      1      0   7.2500         2      2    1   
1         1       1    0  38.0      1      0  71.2833         0      0    2   
2         1       3    0  26.0      0      0   7.9250         2      2    2   
3         1       1    0  35.0      1      0  53.1000         2      0    2   
4         0       3    1  35.0      0      0   8.0500         2      2    1   

   adult_male  embark_town  alive  alone  
0        True            2      0  False  
1       False            0      1  False  
2       False            2      1   True  
3       False            2      1  False  
4        True            2      0   True

Let's see the names of all the columns in the dataset:

In [62]:
# check the names of the columns
print(df.columns)

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'embark_town', 'alive',
       'alone'],
      dtype='object')


In [63]:
# Regression Imputation

# split the data into X and y and we will only take the columns with no missing values
X = df_without_missing.drop(['age'], axis=1)
y = df_without_missing['age']

# split the data into train test sets 
X_train , X_test, y_train , y_test = train_test_split(X, y, test_size=0.25, random_state=42)

# Random Forest Imputation
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

# predict 
y_pred = rf_model.predict(X_test)

# evaluate the model 
print("RMSE for Random Forest Imputation:" , np.sqrt(mean_squared_error(y_test, y_pred)))
print("R2 Score for Random Forest Imputation:" , r2_score(y_test, y_pred))
print("MAE for Random Forest Imputation:" , mean_absolute_error(y_test, y_pred))
print("MAPE for Random Forest Imputation:" , mean_absolute_percentage_error(y_test, y_pred))

RMSE for Random Forest Imputation: 10.518618160957255
R2 Score for Random Forest Imputation: 0.3665795602703329
MAE for Random Forest Imputation: 8.119064807946671
MAPE for Random Forest Imputation: 0.37875369690426836


In [64]:
# check the number of missing values in each column:
df_with_missing.isnull().sum().sort_values(ascending=False)

age            177
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
embarked         0
class            0
who              0
adult_male       0
embark_town      0
alive            0
alone            0
dtype: int64

In [65]:
# predict the missing values
y_pred = rf_model.predict(df_with_missing.drop(['age'], axis=1))

# remove warning
import warnings
warnings.filterwarnings('ignore')

# replace the missing values with the predicted values
df_with_missing['age'] = y_pred

# check the missing values 
df_with_missing.isnull().sum().sort_values(ascending=False)

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

In [66]:
# concatenate the two dataframes 
df_complete = pd.concat([df_with_missing, df_without_missing], axis=0)

# print the shape of the complete dataframe
print("The shape of the complete dataframe is:" , df_complete.shape)

# check the first 5 rows of the complete dataframe 
df_complete.head()



The shape of the complete dataframe is: (891, 14)


survived  pclass  sex        age  sibsp  parch     fare  embarked  class  \
5          0       3    1  33.083958      0      0   8.4583         1      2   
17         1       2    1  36.215384      0      0  13.0000         2      1   
19         1       3    0  17.978333      0      0   7.2250         0      2   
26         0       3    1  36.492542      0      0   7.2250         0      2   
28         1       3    0  20.518333      0      0   7.8792         1      2   

    who  adult_male  embark_town  alive  alone  
5     1        True            1      0   True  
17    1        True            2      1   True  
19    2       False            0      1   True  
26    1        True            0      0   True  
28    2       False            1      1   True

In [67]:
for col in columns_to_encode:
    # retrieve the corresponding LabelEncoder for the column
    le = label_encoders[col]

    # inverse transform the data
    df_complete[col] = le.inverse_transform(df[col])

# check the first 5 rows of the complete dataframe
df_complete.head()

survived  pclass     sex        age  sibsp  parch     fare embarked  \
5          0       3    male  33.083958      0      0   8.4583        S   
17         1       2  female  36.215384      0      0  13.0000        C   
19         1       3  female  17.978333      0      0   7.2250        S   
26         0       3  female  36.492542      0      0   7.2250        S   
28         1       3    male  20.518333      0      0   7.8792        S   

    class    who  adult_male  embark_town alive  alone  
5   Third    man        True  Southampton    no   True  
17  First  woman        True    Cherbourg   yes   True  
19  Third  woman       False  Southampton   yes   True  
26  First  woman        True  Southampton   yes   True  
28  Third    man       False  Southampton    no   True

In [68]:
# print the shape of the complete dataframe
print("The shape of the complete dataframe is:" , df_complete.shape)

The shape of the complete dataframe is: (891, 14)


In [69]:
# check the number of missing values in each column 
df_complete.isnull().sum().sort_values(ascending=False)

embarked       2
embark_town    2
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
class          0
who            0
adult_male     0
alive          0
alone          0
dtype: int64

# 5.Advanced Technique

# 5.1.Multiple Imputation by Chained Equation (MICE)

Let's see how to implement MICE in Python using the Titanic dataset.

In [70]:
# import libraries 
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

# load the dataset
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [71]:
# check the missing values
df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

# SEE

In [85]:
from sklearn.preprocessing import LabelEncoder

# create a LabelEncoder object using LabelEncoder() in for loop for categorical
# columns to encode 
columns_to_encode = ['sex', 'embarked', 'who', 'deck', 'class', 'embark_town', 'age']

# dictionary to store labelencoders for each column
label_encoders = {}

# loop to apply labelencoder to each column
for col in columns_to_encode:
    # create a new labelencoder for each column
    le = LabelEncoder()

    # fit and transform the data, then inverse transform it 
    df[col] = le.fit_transform(df[col])

    # store the encoder in the dictionary 
    label_encoders[col] = le

# check the first few rows of the dataframe
df.head()

survived  pclass  sex  age  sibsp  parch     fare  embarked  class  who  \
0         0       3    1   28      1      0   7.2500         2      2    1   
1         1       1    0   51      1      0  71.2833         0      0    2   
2         1       3    0   34      0      0   7.9250         2      2    2   
3         1       1    0   47      1      0  53.1000         2      0    2   
4         0       3    1   47      0      0   8.0500         2      2    1   

   adult_male  deck  embark_town  alive  alone  
0        True     7            2    0.0  False  
1       False     2            0    1.0  False  
2       False     7            2    1.0   True  
3       False     2            2    1.0  False  
4        True     7            2    0.0   True

In [86]:
# impute the missing values with IterativeImputer 
# call the IterativeImputer class with max_iter = 20
imputer = IterativeImputer(max_iter = 10)

# impute missing values using IterativeImputer in a for loop for age, embark_town, embarked, deck

# columns to impute 
columns_to_impute = ['age', 'embark_town', 'embarked', 'deck']

# loop to impute each column
for col in columns_to_encode:
    df[col] = imputer.fit_transform(df[[col]])
# check the missing values 
df.isnull().sum().sort_values(ascending=False)

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           0
embark_town    0
alive          0
alone          0
dtype: int64

In [87]:
df.head()

survived  pclass  sex   age  sibsp  parch     fare  embarked  class  who  \
0         0       3  1.0  28.0      1      0   7.2500       2.0    2.0  1.0   
1         1       1  0.0  51.0      1      0  71.2833       0.0    0.0  2.0   
2         1       3  0.0  34.0      0      0   7.9250       2.0    2.0  2.0   
3         1       1  0.0  47.0      1      0  53.1000       2.0    0.0  2.0   
4         0       3  1.0  47.0      0      0   8.0500       2.0    2.0  1.0   

   adult_male  deck  embark_town  alive  alone  
0        True   7.0          2.0    0.0  False  
1       False   2.0          0.0    1.0  False  
2       False   7.0          2.0    1.0   True  
3       False   2.0          2.0    1.0  False  
4        True   7.0          2.0    0.0   True

In [89]:
# inverse transform for encoded columns 
for col in columns_to_encode:
    # retrieve the corresponding LabelEncoder for the column
    le = label_encoders[col]
    # inverse transform the data and convert to interger type
    df[col] = le.inverse_transform(df[col].astype(int))

df.head()

survived  pclass  sex   age  sibsp  parch     fare  embarked  class  who  \
0         0       3  1.0  28.0      1      0   7.2500       2.0    2.0  1.0   
1         1       1  0.0  51.0      1      0  71.2833       0.0    0.0  2.0   
2         1       3  0.0  34.0      0      0   7.9250       2.0    2.0  2.0   
3         1       1  0.0  47.0      1      0  53.1000       2.0    0.0  2.0   
4         0       3  1.0  47.0      0      0   8.0500       2.0    2.0  1.0   

   adult_male  deck  embark_town  alive  alone  
0        True   7.0          2.0    0.0  False  
1       False   2.0          0.0    1.0  False  
2       False   7.0          2.0    1.0   True  
3       False   2.0          2.0    1.0  False  
4        True   7.0          2.0    0.0   True